core/src/main/java/org/infinispan/executors/ParallelExecutorHelper.java

```diff
@@ -0,0 +1,59 @@
+package org.infinispan.executors;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Spliterator;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+/**
+ * Static class that provides helpers for parallelizing some operations.
+ *
+ * @author wburns
+ * @since 9.0
+ */
+public class ParallelExecutorHelper {
+   private ParallelExecutorHelper() { }
+
+   /**
+    * Useful to allow for a spliterator to be completely decomposed and submitted to the given Executor.  The results
+    * are then fully consumed by the provided Consumer.  The caller can then be notified of completetion by using the
+    * provided CompletableFuture.
+    * @param spliterator the spliterator to decompose and process each entry
+    * @param consumer the consumer that processes each entry
+    * @param executor the executor to submit the tasks to
+    * @param <T> the type of object being processed
+    * @return future that signifies completion of all submitted tasks
+    */
+   public static <T> CompletableFuture<Void> forEach(Spliterator<? extends T> spliterator,
+           Consumer<? super T> consumer, Executor executor) {
+      long size = spliterator.getExactSizeIfKnown();
+      // If the size is unknown we use the builder method to prevent multiple resizings
+      if (size == -1) {
+         // Use stream builder as it requires less resizing than ArrayList if the spliterator is large
```

Comment:
I don't think keeping all the entries/tasks in memory is a good idea anyway, even if it requires less resizing.
That's the reasoning behind `ExecutorAllCompletionService`, to avoid keeping all the futures in memory.